### PR TITLE
feat: Add struct support to the `Resolver`

### PR DESCRIPTION
**Pull request by dghosef**
This PR is related to issues #7938 and #7936

It adds support to name resolution but no further steps, hence the runtime exceptions I sometimes throw.

For what it's worth, I manually verify all that the `ignore` tests I add would theoretically pass

As a side question: I have been putting all the struct code after the equivalent enum code. Would it make sense to put it after the restrictableenum code? Logically, it might make sense for all the enum constructs to be together and then have the struct ones

### CODE REVIEW

**Comment by magnus-madsen** [comment]
> As a side question: I have been putting all the struct code after the equivalent enum code. Would it make sense to put it after the restrictableenum code? Logically, it might make sense for all the enum constructs to be together and then have the struct ones

I suppose so, but lets deal with it later/separately. RestrictableVariants might be removed.

File: main/src/ca/uwaterloo/flix/language/phase/Resolver.scala [hunk 1]
```diff
@@ -594,6 +612,14 @@ object Resolver {
       }
   }
 
+  private def resolveField(field0: NamedAst.Declaration.StructField, env: ListMap[String, Resolution], taenv: Map[Symbol.TypeAliasSym, ResolvedAst.Declaration.TypeAlias], ns0: Name.NName, root: NamedAst.Root)(implicit flix: Flix): Validation[ResolvedAst.Declaration.StructField, ResolutionError] = field0 match {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Resolver.scala:615]: resolveField -> resolveStructField (just to be precise)
Also, add doc comment.

File: main/src/ca/uwaterloo/flix/language/phase/Kinder.scala [hunk 2]
```diff
@@ -1570,6 +1573,8 @@ object Kinder {
         }
       }
 
+    case UnkindedType.Struct(sym, _) =>
+      throw new RuntimeException("Joe: Structs are not supported in the kind checker yet.")
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/phase/Kinder.scala:1577]: Newline needed :)

File: main/test/ca/uwaterloo/flix/language/phase/TestResolver.scala [hunk 3]
```diff
@@ -89,6 +89,23 @@ class TestResolver extends AnyFunSuite with TestUtils {
...
+         |mod A{
+         |  struct S {
+         |    a: Int32
+         |  }
+         |}
+         |
+         |mod B {
+         |  def g(): A.S = ???
+         |}
+         |"""
```
Comment (magnus-madsen) [inline main/test/ca/uwaterloo/flix/language/phase/TestResolver.scala:104]: We use 4 spaces for indentation :)

Comment (magnus-madsen) [reply main/test/ca/uwaterloo/flix/language/phase/TestResolver.scala:104]: I think one or two more test cases would be welcome.

File: main/src/ca/uwaterloo/flix/language/phase/Resolver.scala (hunk 1 above)
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/phase/Resolver.scala:615]: Needs doc

File: main/src/ca/uwaterloo/flix/language/phase/Resolver.scala [hunk 4]
```diff
@@ -3436,8 +3523,10 @@ object Resolver {
           case _ => Result.Err(ResolutionError.IllegalType(tpe, loc))
         }
 
-      // Case 3: Enum. Return an object type.
+      // Case 3: Enum / Struct. Return an object type.
       case _: UnkindedType.Enum => Result.Ok(Class.forName("java.lang.Object"))
+      // TODO: Should this be a specific class instead
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/phase/Resolver.scala:3528]: I encourage you to use a code name for your TODOs. This makes it easier to find TODOs related to your project.

E.g. `TODO STRUCTS`

**Comment by magnus-madsen** [comment]
@mlutze When satisfied, you can hit merge.

File: main/src/ca/uwaterloo/flix/language/phase/Resolver.scala [hunk 5]
```diff
@@ -3436,8 +3526,10 @@ object Resolver {
           case _ => Result.Err(ResolutionError.IllegalType(tpe, loc))
         }
 
-      // Case 3: Enum. Return an object type.
+      // Case 3: Enum / Struct. Return an object type.
       case _: UnkindedType.Enum => Result.Ok(Class.forName("java.lang.Object"))
+      // TODO STRUCTS Should this be a specific class instead?
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Resolver.scala:3531]: This is OK for now. You can remove the TODO in the next PR

**Review by magnus-madsen — APPROVED** [review]

**Pull request merged**